Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -201,252 +82,618 @@ void LoadSettings()
+					else if (ini.is_value("default_return_system"))
+					{
+						defaultReturnSystem = ini.get_value_string(0);
+					}
+					else if (ini.is_value("default_return_base"))
+					{
+						defaultReturnBase = ini.get_value_string(0);
+					}
+					else if (ini.is_value("target_mobile_docking_base"))
+					{
+						targetMobileDockingBase = CreateID(ini.get_value_string(0));
+					}
+					else if(ini.is_value("disable_docking_repairs"))
+					{
+						disableRegensOnDockAttempt = ini.get_value_bool(0);
+					}
+					else if (ini.is_value("docking_shield_drain"))
+					{
+						disableShieldsOnDockAttempt = ini.get_value_bool(0);
+					}
 				}
 			}
 		}
 		ini.close();
 	}
 
+	if (ini.open(scCarrierDataFile.c_str(), false))
+	{
+		time_t curTime = time(0);
+		while (ini.read_header())
+		{
+			if (ini.is_header("CarrierData"))
+			{
+				CARRIERINFO ci;
+				wstring carrierName;
+				boolean doLoad = true;
+				while (ini.read_value() && doLoad)
+				{
+					if (ini.is_value("carrier"))
+					{
+						carrierName = stows(ini.get_value_string(0));
+						const auto& accData = HkGetAccountByCharname(carrierName);
+						// don't load renamed/deleted ships
+						if (!accData)
+						{
+							doLoad = false;
+						}
+					}
+					else if (ini.is_value("lastLogin"))
+					{
+						if (forgetCarrierDataInSeconds != 0
+							&& (curTime - forgetCarrierDataInSeconds > ini.get_value_int(0))) {
+							doLoad = false;
+						}
+						else {
+							ci.lastCarrierLogin = ini.get_value_int(0);
+						}
+					}
+					else if (ini.is_value("docked"))
+					{
+						wstring& dockedShipName = stows(ini.get_value_string(0));
+						const auto& accData = HkGetAccountByCharname(dockedShipName);
+						// skip loading renamed/deleted docked ships
+						if(!accData){
+							continue;
+						}
+
+						nameToDockedInfoMap[dockedShipName].carrierName = carrierName.c_str();
+						nameToDockedInfoMap[dockedShipName].lastDockedSolar = CreateID(ini.get_value_string(1));
+
+						ci.dockedShipList.push_back(dockedShipName);
+						dockedCount++;
+					}
+				}
+				if (doLoad) {
+					nameToCarrierInfoMap[carrierName] = ci;
+					carrierCount++;
+				}
+				else
+				{
+					for (const auto& dockedShipName : ci.dockedShipList)
+					{
+						MoveOfflineShipToLastDockedSolar(dockedShipName);
+					}
+				}
+			}
+		}
+		ini.close();
+	}
 	ConPrint(L"DockingModules: Loaded %u equipment\n", dockingModAmount);
+	ConPrint(L"DockingModules: Found %u ships docked on %u carriers\n", dockedCount, carrierCount);
 	ConPrint(L"DockingModules: Allowing ships below the cargo capacity of %i to dock\n", cargoCapacityLimit);
 }
 
-void __stdcall BaseExit(uint iBaseID, uint iClientID)
+void SaveData()
 {
-	returncode = DEFAULT_RETURNCODE;
+	if (nameToCarrierInfoMap.empty()){
+		return;
+	}
+	char datapath[MAX_PATH];
+	GetUserDataPath(datapath);
+	string path = string(datapath) + "\\Accts\\MultiPlayer\\docking_module\\mobile_docking.ini";
+	FILE *file = fopen(path.c_str(), "w");

```

Comment:
Correct me if I am wrong, but "w" denotes that it totally replace the existing file. Is this intended?